diag/diag.proto

```diff
@@ -0,0 +1,238 @@
+  // - When no BERT operation, which matches the given BERT operation ID, is in
+  //   progress or completed on any of the ports specified by the request.
+  // - When the BERT operation ID does not match the in progress or completed
+  //   BERT operation on any of the ports specified by the request.
+  // The RPC returns an OK status of none of these situations is encountered.
+  // Note that a BERT operation is considered completed if switch has a
+  // record of it.
+  rpc GetBERTResult(GetBERTResultRequest) returns(GetBERTResultResponse) {}
+
+  // TODO(aghaffar): Add BURNIN related RPCs.
+}
+
+// Common sequence generating monic polynomials used for PRBS.
+enum PrbsPolynomial {
+  PRBS_POLYNOMIAL_UNKNOWN = 0;  // default invalid choice.
+  PRBS_POLYNOMIAL_PRBS7 = 1;
+  PRBS_POLYNOMIAL_PRBS9 = 2;
+  PRBS_POLYNOMIAL_PRBS15 = 3;
+  PRBS_POLYNOMIAL_PRBS20 = 4;
+  PRBS_POLYNOMIAL_PRBS23 = 5;
+  PRBS_POLYNOMIAL_PRBS31 = 6;
+}
+
+// Status returned for each per-port BERT request.
+enum BertStatus {
+  // default invalid choice.
+  BERT_STATUS_UNKNOWN = 0;
+  // BERT requests (Start, Stop, GetStatus) were processed successfully.
+  BERT_STATUS_OK = 1;
+  // The specified port was not found.
+  BERT_STATUS_NON_EXISTENT_PORT = 2;
+  // HW error was encountered while performing BERT operation.
+  BERT_STATUS_HARDWARE_ACCESS_ERROR = 3;
+  // PRBS generating polynomial is not supported by the target.
+  BERT_STATUS_UNSUPPORTED_PRBS_POLYNOMIAL = 4;
+  // There is already a BERT running on the specified port. Returned when
+  // `StartBert` RPC tries to add run BERT on an already in-use port.
+  BERT_STATUS_PORT_ALREADY_IN_BERT = 5;
+  // There is no BERT running on the specified port. Returned when `StopBert`
+  // or `GetBertResult` RPC was called for an idle port.
+  BERT_STATUS_PORT_NOT_RUNNING_BERT = 6;
+  // The specified test duration is too small.
+  BERT_STATUS_TEST_DURATION_TOO_SHORT = 7;
+  // The specified test duration is larger than maximum allowed.
+  BERT_STATUS_TEST_DURATION_TOO_LONG = 8;
+  // The given BERT operation ID is not known. Returned for `StopBert` and
+  // `GetBertResult` RPCs.
+  BERT_STATUS_BERT_OPERATION_ID_NOT_FOUND = 9;
+  // The given BERT operation ID is already in use. Returned when `StartBert`
+  // RPC uses an ID which is already memorized for a BERT operation.
+  BERT_STATUS_BERT_OPERATION_ID_ALREADY_IN_USE = 10;
+  // Failure to get the peer lock.
+  BERT_STATUS_PEER_LOCK_FAILURE = 11;
+  // Lost the peer lock after locking once.
+  BERT_STATUS_PEER_LOCK_LOST = 12;
+  // Misc internal errors that cannot be categorized by any of the previous
+  // error codes.
+  BERT_STATUS_INTERNAL_ERROR = 13;
+}
+
+message StartBERTRequest {
+  // Per port BERT start requests.
+  message PerPortRequest {
+    // Path to the interface corresponding to the port.
+    gnoi.Path interface = 1;  // required
+    // The selected PRBS generating polynomial for BERT.
+    PrbsPolynomial prbs_polynomial = 2;  // required
```

Comment:
Is it required to be able to run different polynomials per port? I'm not clear what the use case for this would be.

Reply:
Chips support this. I see no harm in allowing this.